cardano-cli/src/Cardano/CLI/Plutus/Minting.hs

```diff
@@ -0,0 +1,200 @@
+{-# LANGUAGE DataKinds #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE LambdaCase #-}
+
+module Cardano.CLI.Plutus.Minting
+  ( CliMintScriptRequirements (..)
+  , MintScriptWitWithPolId (..)
+  , createOnDiskSimpleOfPlutusScriptCliArgs
+  , createOnDiskSimpleReferenceScriptCliArgs
+  , createOnDiskPlutusReferenceScriptCliArgs
+  , CliScriptWitnessError
+  , readMintScriptWitness
+  )
+where
+
+import           Cardano.Api
+import           Cardano.Api.Shelley
+
+import           Cardano.CLI.Read
+import           Cardano.CLI.Types.Common (ScriptDataOrFile)
+
+-- We always need the policy id when constructing a transaction that mints.
+-- In the case of reference scripts, the user currently must provide the policy id (script hash)
+-- in order to correctly construct the transaction.
+data MintScriptWitWithPolId era
+  = MintScriptWitWithPolId
+  { mswPolId :: PolicyId
+  , mswScriptWitness :: ScriptWitness WitCtxMint era
+  }
+  deriving Show
+
+data OnDiskSimpleOrPlutusScriptCliArgs
+  = OnDiskSimpleScriptCliArgs
+      (File ScriptInAnyLang In)
+  | OnDiskPlutusScriptCliArgs
+      (File ScriptInAnyLang In)
+      ScriptDataOrFile
+      ExecutionUnits
+  deriving Show
+
+createOnDiskSimpleOfPlutusScriptCliArgs
+  :: File ScriptInAnyLang In
+  -> Maybe (ScriptDataOrFile, ExecutionUnits)
+  -> CliMintScriptRequirements
+createOnDiskSimpleOfPlutusScriptCliArgs scriptFp Nothing =
+  OnDiskSimpleOrPlutusScript $ OnDiskSimpleScriptCliArgs scriptFp
+createOnDiskSimpleOfPlutusScriptCliArgs scriptFp (Just (redeemerFile, execUnits)) =
+  OnDiskSimpleOrPlutusScript $ OnDiskPlutusScriptCliArgs scriptFp redeemerFile execUnits
+
+data SimpleRefScriptCliArgs
+  = SimpleRefScriptCliArgs
+      TxIn
+      PolicyId
+  deriving Show
+
+createOnDiskSimpleReferenceScriptCliArgs
+  :: TxIn
+  -> PolicyId
+  -> CliMintScriptRequirements
+createOnDiskSimpleReferenceScriptCliArgs txin polid =
+  OnDiskSimpleRefScript $ SimpleRefScriptCliArgs txin polid
+
+data PlutusRefScriptCliArgs
+  = PlutusRefScriptCliArgs
+      TxIn
+      AnyPlutusScriptVersion
+      ScriptDataOrFile
+      ExecutionUnits
+      PolicyId
+  deriving Show
+
+createOnDiskPlutusReferenceScriptCliArgs
+  :: TxIn
+  -> AnyPlutusScriptVersion
+  -> ScriptDataOrFile
+  -> ExecutionUnits
+  -> PolicyId
+  -> CliMintScriptRequirements
+createOnDiskPlutusReferenceScriptCliArgs txin scriptVersion scriptData execUnits polid =
+  OnDiskPlutusRefScript $ PlutusRefScriptCliArgs txin scriptVersion scriptData execUnits polid
+
+data CliMintScriptRequirements
+  = OnDiskSimpleOrPlutusScript OnDiskSimpleOrPlutusScriptCliArgs
+  | OnDiskSimpleRefScript SimpleRefScriptCliArgs
+  | OnDiskPlutusRefScript PlutusRefScriptCliArgs
+  deriving Show
+
```

Comment:
Same like my previous comment, can you put this definition near the top of the module, before any of its "dependencies"?

(also applies to other types here, I don't want to highlight the whole file for that purpose)